src/async/async.c

```diff
@@ -351,7 +351,8 @@ void re_async_cancel(struct re_async *async, intptr_t id)
 		w->workh = NULL;
 		w->cb	 = NULL;
 		w->arg	 = mem_deref(w->arg);
-		list_move(&w->le, &async->freel);
+		/* No move to free list since queueh must always handled if
+		 * mqueue_push is called */
```

Comment:
I can't see why this change has something to do if the pipe is read or not.

Reply:
Pipe queueh is always executed (since it's FIFO), so the free list move must be done there.